sdk/src/main/scala/io/horizen/account/state/ForgerStakeV2MsgProcessor.scala

```diff
@@ -49,30 +51,87 @@ object ForgerStakeV2MsgProcessor extends NativeSmartContractWithFork  with Forge
     }
   }
 
-  def doDelegateCmd(invocation: Invocation, gasView: BaseAccountStateView, msg: Message): Array[Byte] = {
+  def doDelegateCmd(invocation: Invocation, view: BaseAccountStateView, context: ExecutionContext): Array[Byte] = {
+
+    checkForgerStakesV2IsActive(view)
+
     val inputParams = getArgumentsFromData(invocation.input)
-    val cmdInput = DelegateCmdInputDecoder.decode(inputParams)
-    log.info(s"delegate called - ${cmdInput.forgerPublicKeys}")
-    //TODO: add logic
+    val DelegateCmdInput(forgerPublicKeys) = DelegateCmdInputDecoder.decode(inputParams)
+
+    log.info(s"delegate called - $forgerPublicKeys")
+    val stakedAmount = invocation.value
+
+    if (stakedAmount.signum() <= 0) {
+      throw new ExecutionRevertedException("Value must not be zero")
+    }
+
+    if (!isValidZenAmount(stakedAmount)) {
+      throw new ExecutionRevertedException(s"Value is not a legal wei amount: $stakedAmount")
+    }
+
+    if (view.getBalance(invocation.caller).compareTo(stakedAmount) < 0){
+      throw new ExecutionRevertedException(s"Insufficient funds. Required: $stakedAmount, available: ${view.getBalance(invocation.caller)}")
+    }
+
+    val epochNumber = context.blockContext.consensusEpochNumber
+
+    StakeStorage.addStake(view, forgerPublicKeys.blockSignPublicKey, forgerPublicKeys.vrfPublicKey,
+      epochNumber, invocation.caller, stakedAmount)
+
+    val delegateStakeEvt = DelegateForgerStake(invocation.caller, forgerPublicKeys.blockSignPublicKey, forgerPublicKeys.vrfPublicKey, stakedAmount)
+    val evmLog = getEthereumConsensusDataLog(delegateStakeEvt)
+    view.addLog(evmLog)
+
+    view.subBalance(invocation.caller, stakedAmount)
+    // increase the balance of the "forger stake smart contract” account
+    view.addBalance(contractAddress, stakedAmount)
+
+
     Array.emptyByteArray
   }
 
-  def doWithdrawCmd(invocation: Invocation, gasView: BaseAccountStateView, msg: Message): Array[Byte] = {
+  def doWithdrawCmd(invocation: Invocation, view: BaseAccountStateView, context: ExecutionContext): Array[Byte] = {
+    requireIsNotPayable(invocation)
+    checkForgerStakesV2IsActive(view)
+
     val inputParams = getArgumentsFromData(invocation.input)
-    val cmdInput = WithdrawCmdInputDecoder.decode(inputParams)
-    log.info(s"withdraw called - ${cmdInput.forgerPublicKeys} ${cmdInput.value}")
+    val WithdrawCmdInput(forgerPublicKeys, stakedAmount) = WithdrawCmdInputDecoder.decode(inputParams)
+
+    if (stakedAmount.signum() != 1) throw new ExecutionRevertedException(s"Withdrawal amount must be greater than zero: $stakedAmount")
+
+    if (!isValidZenAmount(stakedAmount)) {
+      throw new ExecutionRevertedException(s"Value is not a legal wei amount: $stakedAmount")
+    }
+
+    log.debug(s"withdraw called - $forgerPublicKeys $stakedAmount")
```

Comment:
I would move this at line 99. Moreover, consider adding info/warn trace in case of exception throwing, that can be useful especially in early dev phase.